### DISCUSSION TITLE
[06/08] 피어세션 일지

### DISCUSSION THREAD

**Discussion by ddoyoon** [bcaitech1/p4-dkt-lastffang#103] [Peer Session]
### 브리핑

- @ddoyoon 
    - Stratified k-fold 구현: data augmentation 과 잘 맞지 않는 부분이 있어서 뭘 기준으로 데이터 나눌지 좀 실험해봐야 함. 시간이 걸릴 것 같아서 우선 augmentation 없는 현재 마스터 기준으로 작업하고 데이터 비교 실험
    - lstmattn 모델에 대해서는 수정한 데이터가 약간 더 성능이 좋은데 last query 모델에 대해 확인하고 드라이브 링크로 데이터 드릴 예정
    - 진홍님 언급하신 elapsed time 쓰는 코드: 간단히 column 존재 여부 확인하는거로.. 오늘 중으로 pr 날리겠음
    - 건희님 발견해주신 cv inference 버그는 모델 정의 전에 seed 한번 고정해주면 해결되는듯
    - 윤석님 언급하신 cv 옵션 argument 추가: cv 안 할 때는 더더욱 train-validation 이 고른 분포여야 할 것 같아서, 그냥 skf 에서 하나의 fold 만 하는 식으로
 
- @scarlettJin 
    - 서버 바꾸면서 베이스 모델 성능이 달라져서 테스트 다시 하고 베이스라인 잡을 예정
        - 새봄: 간단한 피쳐는 LGBM 에서 확인하는게 빠를수도 있다
    - 서빙 강의 수강

- @alinghi
    - 파이캐럿 model search 돌려보는데 모델이 커서 그런지 아직 안끝남 (60시간 예상)
    - LGBM 성능이 가장 좋다. Gradient boost, Adaboost 도 괜찮음
    - 피쳐를 많이 넣으면 메모리 터져서 개수 줄였고 전부 numerical 로 넣었음 ACC 0.76~77 정도
        - 새봄: 서일님 feature + knowledgetag 정도만 넣기 추천

- @gonipark 
    - AUC average wandb 에 올리는거 진행중 
  
- @newspring97
    - riiid 데이터 결론: 못 쓸 것 같다 :(
        - 한 task container 안에 문제가 적은편 (보통 1, 많으면 3문제). testID 로 완전 치환하기는 어려워보인다
        - task container 마다 numbering 이 user 시작 시점이다 (task container id 가 같아도 우리 데이터처럼 같은 시험지/같은 문제 아님)
        - 데이터 포팅은 일단 중단
    - 오늘것까지 서빙 강의, 실습 들었는데 airflow 쓰는 부분 다시 듣고 이해 필요하다

- @GY-Jeong 
    - 모델 임베딩 여러가지 시도하는데 아직은 현재 방식이 제일 잘됨
    - 랜덤성 관련 성능 달라지는 부분 디버깅: dataloader 만들어질 때 batch 순서가 다르게 나옴
        - 일단 해결 방법은 shuffle 끄는것. 순서가 동일하게 들어가는 문제는 k-fold 하면서 해결될 것
    - LayerNorm 일단 사용 보류 (branch 는 있고 PR 은 닫음)
        - lstm 에서는 예상처럼 작동했는데, lstmattn 모델에서 fold 마다 적용했을 때 적용 안한게 validation 이 훨씬 좋았음
    - 전체 train 으로 학습하기 → 제출하니 성능이 별로 안 좋았다

### 서빙
- 어떤 데이터를 집어넣으면 좋을지?
- 서빙은 추천시스템 계열로 가야할 것 같다는 새봄님 의견 (e.g. 영화 좋았다/싫었다 yes or no..)
    - KnowledgeTag ~= 영화장르.. 대체하기 어려운 feature 는 생략하는 방향으로 모델 수정하기?
    - 문제 데이터가 없어서 우리 플젝의 presentation 으로 쓰기는 어려울 것도 같다
- 어쩌면 MRC 데이터로 서빙하는 방향으로 틀 수도 있다.
    - 새봄님 팀 - T5 generation 모델 다시 써보기?
- 네트워킹데이 때 발표할건 아니지만 github 에 보여줄 수 있는 형태 / 데모 영상은 나왔음 좋겠다.
- 시작 기간은 언제쯤?
    - 등수 약간 포기하고 일찍 시작하기 아님 네트워킹데이 이후에 시작하기?
- **서빙 관해서는 금요일에 다시 얘기하자**

### 데이터
- 영화 데이터도 개봉일자 기준으로 시계열로 처리해서 핫한 영화를 뽑는 알고리즘 있었다. 각 유저 맨 뒤 20개가 아니라 DGIM 샘플링으로 뽑은 20개 데이터를 입력해보기?
### 앙상블
- hard voting?  Step function 형태로 되니까 AUROC 측면에서는 효과가 별로.
- 구간별로

### 멘토링 (질문 답변)
- tf 구현체를 다른 framework로?
    - 왕도/방법이 없다..! 어쩔 수 없이 두개의 framework 둘 다 구현 가능해야 재구현 가능할듯. seq2seq 등 유명한 모델 직접 짜면서 두 framework 비교
    - 논문에서 리포트된 성능이 나오는지로 구현을 평가
- inference 도 random seed 영향?
    - 베이스라인에서 임베딩만 바꾼건데 랜덤성이 생긴 이유는 직관적으로는 조교님도 잘 모르겠다. 모델을 불러올 때 랜덤성이 있긴 하다 (e.g. VAE 같은데서 gaussian sampling inference 나 text generation 에서 생성하면서 top-k 할 때..?) 근데 이 task 에 대해서는 잘 모르겠다.
- 랜덤성 관련 추가질문: 각각 다른 서버마다 동일 실험이 다른 그래프 그려지는 문제?
    - 일단 하드웨어가 동일해야 함. 하드웨어마다 불러오는 랜덤 시드 테이블이 다름
    - 쿠다, 파이썬, 파이토치 등 버전 전부 동일하게 하고 실험해야함
    - 같은 GPU 상에서 실험하는게 해석하기 더 좋을 것이다
- (미해결) Augmentation
    - 논문(task 에 적용 가능할지는 불확실): [Time Series Data Augmentation for Deep Learning: A Survey](https://arxiv.org/abs/2002.12478)
        - 1, 2 번은 저렇게 해도 되나 싶긴 한데 3번부터는 NMT 등에서 많이 쓰이는 방법
    - Augmentation 에 대해서는 task 에 대해 좀 더 보고 다시 알려주실 것
- 정형 데이터 형태인데 주어지는 feature 가 많지 않아 나름대로 추가해 썼는데 다른 접근법?
    - 사용자별로 과거 정답률로 실력을 가늠할 수 있게?
        - e.g. 수능 30번 맞은 사람들은 대체로 27~29 도 맞았을듯
    - sequence 를 연달아 쓰는게 아니라 ~20번까지만 보고 30번의 정답 여부를 보는 것 등. 샘플을 부분부분 보는 것..? Auxiliary task처럼 학습하면 data augmentation 의 효과도 낼 수 있을 것 같다.
        - t 까지의 데이터로 t+1 ~ t+n 까지 예측하게 한다
            - 자연어 생성: 사람들이 한 단어씩 생각해서 말하지 않는다는 intuition 으로 모델링
- (미해결) 모델 서빙 관련: 지금 하고 있는게 일반적인 task 가 아니라 어떤 방향으로 해야할지 모르겠다. 예시나 경험?
    - 사전에 10문제를 풀었을 때 모의고사 점수 예측하기?
        - 문제 데이터가 주어지지 않아서 어렵다
    - 좀 더 생각해보고 답변 주실 것
- 조교님이 하고 계신 연구 분야: Text generation, interactive NMT (사용자 입력을 반영해서 번역하기)
    - 요즘의 paradigm: 자연어는 데이터가 많으면 다 풀 수 있다는 쪽으로 가는데 관심 가는 영역은 모델의 generalization.
    - 보통은 linear transformation 으로 class 를 맞히도록 하는데, 이제는 모든 것을 다 generation 으로 풀도록 모델을 학습한다.
        - e.g. 5지선다에서 6번까지 있는 문제 → 전자는 재학습해야 함
           - ODQA 도 entity generation 방식으로 풀려는 접근이 있다.
    - Generation model 은 어떤 것? 자기가 풀려는 문제에 따라 BART 를 task 에 맞게 제약조건을 걸어서 쓴다.
    - Naver AI deview (NAVER AI NOW : https://naver-ai-now.kr/)? 보길 권장
    - 다음 미팅은 금요일 4:30

**Comment by GY-Jeong** [comment]
한줄 수정했습니당
Naver AI deview (NAVER AI NOW : https://naver-ai-now.kr/)?
저는 이렇게 들었는데 최근에 네이버에서 GPT-3와 같은 아주 큰 모델(하이퍼클로바..?)을 한국어에 적용한 방법들을 발표했는데, 그 내용에서 언어모델을 어떻게 학습시켰고 데이터를 어떻게 수집했는지 내용이 있습니다
